### PR TITLE
refactor: std/math/nonnative -> std/math/emulated

### PR DESCRIPTION
**Pull request by gbotrel**
This PR refactors std/math/nonnative -> std/math/emulated. 

`emulated.Field` has no pointer receiver operations, and only usage is through `NewField` . 
Emulated arithmetic is now parametrized with the emulated field constants; usage is changed to: 

```
type Circuit struct {
	// Limbs of non-native elements X, Y and Res
	X, Y, Res emulated.Element[emulated.Secp256k1]
}

func (circuit *Circuit) Define(api frontend.API) error {
	// wrap API to work in SECP256k1 scalar field
	secp256k1, err := emulated.NewField[emulated.Secp256k1](api)
	if err != nil {
		return err
	}

	tmp := secp256k1.Mul(circuit.X, circuit.Y)
	secp256k1.AssertIsEqual(tmp, circuit.Res)
	return nil
}

// test file:
func TestEmulatedArithmetic(t *testing.T) {
	assert := test.NewAssert(t)
	std.RegisterHints()

	var circuit, witness Circuit

	witness.X.Assign("26959946673427741531515197488526605382048662297355296634326893985793")
	witness.Y.Assign("53919893346855483063030394977053210764097324594710593268653787971586")
	witness.Res.Assign("485279052387156144224396168012515269674445015885648619762653195154800")

	assert.ProverSucceeded(&circuit, &witness, test.WithCurves(ecc.BN254), test.WithBackends(backend.GROTH16), test.NoSerialization())
}

```

### CODE REVIEW

**Comment by gbotrel** [comment]
@ivokub I think we should look at merging this as is -- this PR doesn't do any algorithmic change on your work, just cosmetic / refactoring. 
A future PR will be needed to improve perf 👍 

Will also add a warning that this feature is "experimental".

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/hints.go b/std/math/emulated/hints.go
index da27bb1f..557b464a 100644
--- a/std/math/emulated/hints.go
+++ b/std/math/emulated/hints.go
@@ -12,10 +12,7 @@ import (
 // inside a func, then it becomes anonymous and hint identification is screwed.
 
 func init() {
-	hints := GetHints()
-	for _, h := range hints {
-		hint.Register(h)
-	}
+	hint.Register(GetHints()...)
 }
 
 // GetHints returns all hint functions used in the package.

```

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/params.go b/std/math/emulated/params.go
index fce8a1c3..067b6c9f 100644
--- a/std/math/emulated/params.go
+++ b/std/math/emulated/params.go
@@ -51,7 +51,7 @@ func (fp BN254Fp) BitsPerLimb() uint { return 32 }
 func (fp BN254Fp) IsPrime() bool     { return true }
 func (fp BN254Fp) Modulus() *big.Int { return ecc.BN254.BaseField() }
 
-// BLS12377Fp provide type parametrization for emulated field on 8 limb of width 32bits
+// BLS12377Fp provide type parametrization for emulated field on 6 limb of width 64bits
 // for modulus 0x1ae3a4617c510eac63b05c06ca1493b1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
 type BLS12377Fp struct{}
 

```

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/field.go b/std/math/emulated/field.go
index e3f1f533..5e2cdd6f 100644
--- a/std/math/emulated/field.go
+++ b/std/math/emulated/field.go
@@ -16,10 +16,6 @@ import (
 	"github.com/rs/zerolog"
 )
 
-type API interface {
-	frontend.API
-}
-
 // field defines the parameters of the emulated ring of integers modulo n. If
 // n is prime, then the ring is also a finite field where inverse and division
 // are allowed.

```

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/field_test.go b/std/math/emulated/field_test.go
index 235a583c..06bece44 100644
--- a/std/math/emulated/field_test.go
+++ b/std/math/emulated/field_test.go
@@ -220,6 +220,7 @@ func (circuit *pairingBLS377) Define(api frontend.API) error {
 }
 
 func TestPairingBLS377(t *testing.T) {
+	t.Skip()
 	assert := test.NewAssert(t)
 
 	_, _, P, Q := bls12377.Generators()
diff --git a/std/math/emulated/pairing_test.go b/std/math/emulated/pairing_test.go
index cec46b2c..c7b722c1 100644
--- a/std/math/emulated/pairing_test.go
+++ b/std/math/emulated/pairing_test.go
@@ -20,6 +20,7 @@ func (circuit *mlBLS377) Define(api frontend.API) error {
 }
 
 func TestE12SquareBLS377(t *testing.T) {
+	t.Skip()
 	assert := test.NewAssert(t)
 
 	circuit := mlBLS377{

```

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/element.go b/std/math/emulated/element.go
index 180b94dd..59747528 100644
--- a/std/math/emulated/element.go
+++ b/std/math/emulated/element.go
@@ -410,9 +410,6 @@ func (f *field[T]) assertIsEqual(a, b Element[T]) Element[T] {
 	// so essentially, we say "I know an element k such that k*p == diff"
 	// hence, diff == 0 mod p
 	p := f.Modulus()
-	// we compute k such that diff / p == k
-	// so essentially, we say "I know an element k such that k*p == diff"
-	// hence, diff == 0 mod p
 	k, err := f.computeQuoHint(diff)
 	if err != nil {
 		panic(fmt.Sprintf("hint error: %v", err))

```

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/element.go b/std/math/emulated/element.go
index 180b94dd..78c48ced 100644
--- a/std/math/emulated/element.go
+++ b/std/math/emulated/element.go
@@ -551,10 +551,3 @@ func max[T constraints.Ordered](a, b T) T {
 	}
 	return b
 }
-
-func min[T constraints.Ordered](a, b T) T {
-	if a < b {
-		return a
-	}
-	return b
-}

```

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/hints.go b/std/math/emulated/hints.go
index da27bb1f..4276771d 100644
--- a/std/math/emulated/hints.go
+++ b/std/math/emulated/hints.go
@@ -110,9 +110,7 @@ func RemHint(_ *big.Int, inputs []*big.Int, outputs []*big.Int) error {
 	if err != nil {
 		return err
 	}
-	for _, o := range outputs {
-		o.SetUint64(0)
-	}
+
 	r := new(big.Int)
 	r.Rem(x, y)
 	if err := decompose(r, nbBits, outputs); err != nil {

```

File: std/math/emulated/field_test.go [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+	keys := make([]string, 0, len(circuits.Circuits))
+	for k := range circuits.Circuits {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	for i := range keys {
+		name := keys[i]
+		if name == "inv" || name == "div" || name == "cmp" {
+			// TODO @gbotrel thes don't pass when we use emulated field modulus != snark field
```
Comment (ivokub) [inline std/math/emulated/field_test.go:147]: Hmm. I was at some point fixing these failing tests. I'll try to pop the stash. I think inv and div were easy but there was some strange problem with cmp. I'll  see.

File: std/math/emulated/field_test.go [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+			assert.Run(func(assert *test.Assert) {
+				err := test.IsSolved(tData.Circuit, assignment, testCurve.ScalarField(), wrapperOpt)
+				assert.Error(err)
+			}, name, fmt.Sprintf("invalid=%d", i))
+		}
+	}
+}
+
+func TestVarToElements(t *testing.T) {
+	assert := require.New(t)
```
Comment (ivokub) [inline std/math/emulated/field_test.go:173]: We're using wrapped assert everywhere else.

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/math/emulated/field_test.go b/std/math/emulated/field_test.go
index 235a583c..991193f9 100644
--- a/std/math/emulated/field_test.go
+++ b/std/math/emulated/field_test.go
@@ -15,7 +15,6 @@ import (
 	"github.com/consensys/gnark/std/algebra/fields_bls12377"
 	"github.com/consensys/gnark/std/algebra/sw_bls12377"
 	"github.com/consensys/gnark/test"
-	"github.com/stretchr/testify/require"
 )
 
 func witnessData(q *big.Int) (X1, X2, X3, X4, X5, X6, Res *big.Int) {
@@ -170,7 +169,7 @@ func TestIntegrationApi(t *testing.T) {
 }
 
 func TestVarToElements(t *testing.T) {
-	assert := require.New(t)
+	assert := test.NewAssert(t)
 	_f, _ := NewField[BN254Fp](nil)
 
 	f := _f.(*field[BN254Fp])

```

File: std/math/emulated/element.go [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+	}
+}
+
+// Set sets e to a and returns e. If a is constant, then it also enforces the
+// widths of the limbs.
+func (e *Element[T]) Set(a Element[T]) {
+	e.Limbs = make([]frontend.Variable, len(a.Limbs))
+	e.overflow = a.overflow
+	copy(e.Limbs, a.Limbs)
+	// TODO @gbotrel this shouldn't happen anymore
```
[the remark this answers is not in the document]
Comment (gbotrel) [reply std/math/emulated/element.go:386]: good point. From the api usage, it's not an issue (witness will use NewElement or Assign). But techniically, the limbs are not constrained. Will track in a separate issue, not sure it's a quick fix.

[the remark this answers is not in the document]
Comment (gbotrel) [reply std/math/emulated/element.go:386]: see [#348](https://github.com/ConsenSys/gnark/issues/348)

File: std/math/emulated/hints.go [hunk 4]
```diff
@@ -75,117 +85,96 @@ func MultiplicationHint(mod *big.Int, inputs []*big.Int, outputs []*big.Int) err
...
-		params.nbLimbs,
+		fp.BitsPerLimb(),
+		len(x.Limbs),
 	}
-	p := params.Modulus()
-	for i := range p.Limbs {
-		hintInputs = append(hintInputs, frontend.Variable(p.Limbs[i]))
+	hintInputs = append(hintInputs, x.Limbs...)
+	hintInputs = append(hintInputs, y.Limbs...)
+	limbs, err := f.api.NewHint(RemHint, int(len(y.Limbs)), hintInputs...)
```
[the remark this answers is not in the document]
Comment (gbotrel) [reply std/math/emulated/hints.go:98]: yes that's the point? so the result size is bounded by y-1 ?

File: std/math/emulated/params.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+// BN254Fp provide type parametrization for emulated field on 8 limb of width 32bits
+// for modulus 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47
+type BN254Fp struct{}
+
+func (fp BN254Fp) NbLimbs() uint     { return 8 }
+func (fp BN254Fp) BitsPerLimb() uint { return 32 }
+func (fp BN254Fp) IsPrime() bool     { return true }
+func (fp BN254Fp) Modulus() *big.Int { return ecc.BN254.BaseField() }
+
+// BLS12377Fp provide type parametrization for emulated field on 8 limb of width 32bits
```
[the remark this answers is not in the document]
Comment (gbotrel) [reply std/math/emulated/params.go:54]: well, it tests more code path for one. But also mostly, no good reasons; was experimenting with pairing emulation.

File: std/math/emulated/hints.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivokub) [reply std/math/emulated/hints.go:98]: Yes, it works as a general-purpose method (than can be used for any `y`, not only for `y=modulus`). But in practice use only for `y=modulus`. In that case maybe can remove `y` from the arguments and take modulus from the type parameter? But it also works currently as-is.

File: std/math/emulated/params.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivokub) [reply std/math/emulated/params.go]: Yes, makes sense to test more paths. When debugging I discovered quite a few bugs when had edge cases (limbs fully divided scalar field width, upper limb had some extreme number of bits etc. etc.).

**Comment by gbotrel** [comment]
>Looks good. Does it make sense to separate limb width into second type parameter (instead of embedding it in FieldParams)?

Well, I see the point. One one hand, it hurts readability, but IF we need to emulate same field with different parameters for limbs, it's cleaner. 

I suspect circuit developer will have their own struct (fieldParams) anyhow to tune that depending on target...?

**Comment by ivokub** [comment]
> > Looks good. Does it make sense to separate limb width into second type parameter (instead of embedding it in FieldParams)?
> 
> Well, I see the point. One one hand, it hurts readability, but IF we need to emulate same field with different parameters for limbs, it's cleaner.
> 
> I suspect circuit developer will have their own struct (fieldParams) anyhow to tune that depending on target...?

I think the circuit developers would in general be interested in some standard fields (maybe one for every fp/fr in gnark-crypto), not really caring about rest (limb widths etc.). If there is some special use case, then the developer can define its own type with their own modulus and limb size.

I also think that we really cannot optimise easily th width as the final number of constraints depends on the non-native operations. (Probably to obtain a good heuristic we should just compile with different `nbBits` and find the value which minimises the number of constraints). So it would be good to have some sane defaults, for which the current ones make sense.
